### DISCUSSION TITLE
Whats the difference between /Applications/Emacs.app/Contents/MacOS/Emacs and /usr/local/bin/emacs?

### DISCUSSION THREAD

**Discussion by gwbrck** [d12frosted/homebrew-emacs-plus#557] [Q&A]
... and why does `pasword-store-edit`gets closed immediately with emacs, but works as intended with  /Applications/Emacs.app/Contents/MacOS/Emacs? 

Password-store-edit relies on `with-editor-asynv-shell-command`. So, something semms to be off with that in one of the emacs "versions". 

Please help me understand... Thank you!

**Comment by d12frosted** [comment]
Hey,

The `emacs` binary probided by Emacs+ is just a shortcut for running `Emacs.app`. You can verify that by checking it's content:

```bash
#!/bin/bash
exec /opt/homebrew/Cellar/emacs-plus@30/30.0.50/Emacs.app/Contents/MacOS/Emacs "$@"
```

What is `password-store-edit`? How do you run it?

> works as intended with /Applications/Emacs.app/Contents/MacOS/Emacs

I think it's worth asking - how do you start Emacs? I mean, fundamentally there are two ways (though there are quirky details). I might use incorrect dictionary, but basically you can either run the application via launchd (e.g. finder, spotlight, etc.) or via command line in your interactive shell. The latter can be done by simply calling `/opt/homebrew/Cellar/emacs-plus@30/30.0.50/Emacs.app/Contents/MacOS/Emacs` from your terminal or by calling `emacs` binary. But again, details are what matter :)

**Reply by gwbrck** [reply]
Thanks for your answer, and thank you for guiding my question in the right direction.

So, first of all `password-store-edit` is a function in [password-store.el](https://git.zx2c4.com/password-store/tree/contrib/emacs/password-store.el ) which is a wrapper around the [pass commandline program](https://www.passwordstore.org/). The wrapper uses `with-editor-async-shell-command` and calls `pass edit`. Pass decrypts a password file and opens it with $EDITOR for edit. With`with-editor-async-shell-command` the $EDITOR var gets set to the current instance of Emacs (At least that's my current understanding...). As described above, when I start emacs with `emacs` in fish-shell than the async-process the process finishes immediately without prompting the decrypted password file. The strange thing is that it works when I start emacs with the following prompt in fish: `/Applications/Emacs.app/Contents/MacOS/Emacs`. In this case, `password-store-edit`I get two buffers. One with *Async Shell Output* (it states, "Waiting for Emacs...") and one with the decrypted password file. So, I guess this is how it should be.

So yes, you are probably right. It's about how I start emacs. But actually, I am a bit lost.
Normally, I start Emacs from within fish with `emacsclient -cn -a ""`

**Reply by gwbrck** [reply]
Investigating it further, it boils down to:
`/usr/local/opt/emacs-plus@29/Emacs.app/Contents/MacOS/Emacs`working and 
`/usr/local/Cellar/emacs-plus@29/29.0.60/Emacs.app/Contents/MacOS/Emacs`not working. 
The funny thing is that the former is only a symlink to the latter.

Any ideas? 
Thank you!

**Reply by gwbrck** [reply]
Investigating it even further, I found that the difference is in how the var `with-editor-emacsclient-executable` gets set. In the "working case" it gets set to `/usr/local/bin/emacsclient`. In the "not working case" it gets set to `/usr/local/Cellar/emacs-plus@29/29.0.60/bin/emacsclient`. Independent of how I start emacs setting `with-editor-emacsclient-executable` to `/usr/local/bin/emacsclient` always works.

**Reply by d12frosted** [reply]
Are you sure Emacs server is not interfering somehow? Can you also try testing it all with `-Q` option? E.g. `emacs -Q`

**Reply by gwbrck** [reply]
Yes, I tried that. Maybe it's because `with-editor` requires `'server`? So, perhaps it is a `with-editor`-package issue.
I have this in my config now:
```
(setq with-editor-emacsclient-executable (executable-find "emacsclient"))
```
It's working, but I have the feeling that something is misconfigured on my side...

**Reply by d12frosted** [reply]
I understand the difference between running Emacs.app from Finder/Spotlight and running `emacs` binary from terminal. This difference is a reason behind [PATH injection](https://d12frosted.io/posts/2022-05-12-emacs-plus-path-injection.html) in Emacs+. But I don't understand why you get different experience between `/usr/local/opt/emacs-plus@29/Emacs.app/Contents/MacOS/Emacs`  and `/usr/local/Cellar/emacs-plus@29/29.0.60/Emacs.app/Contents/MacOS/Emacs`. On the other hand, I don't have homebrew stuff under `/user/local` cause I am not on Intel 🤷 So can't really test it.

BTW, maybe try with Emacs 30? So you can simply `brew link` between versions.
